src/routing_node.rs

```diff
@@ -836,6 +842,64 @@ impl RoutingNode {
                                   refresh_our_own_group: bool) -> RoutingResult {
         unimplemented!()
     }
+
+    fn handle_cache_get(&mut self, request : ExternalRequest,
+                                   from_authority : Authority, response_token : SignedToken)
+         -> RoutingResult {
+        let response_data = match request {
+            ExternalRequest::Get(data_request) => {
+                match data_request {
+                    DataRequest::StructuredData(_, _) =>
+                        return Err(From::from(ResponseError::NoData)),
+                    DataRequest::ImmutableData(data_name, _) => {
+                        match self.immutable_data_cache.get(&data_name) {
+                            Some(ref data) => (*data).clone(),
+                            _ => return Err(From::from(ResponseError::NoData)),
+                        }
+                    }
+                    DataRequest::PlainData(data_name) =>
+                    match self.plain_data_cache.get(&data_name) {
+                        Some(ref data) => (*data).clone(),
+                        _ => return Err(From::from(ResponseError::NoData)),
+                    }
+                }
+            },
+            _ => return Err(From::from(ResponseError::NoData)),
+        };
+
+        let routing_message = RoutingMessage {
+            from_authority : Authority::ManagedNode(self.core.id().name()),
+            to_authority   : Authority::ManagedNode(from_authority.get_location().clone()),
+            content        : Content::ExternalResponse(
+                ExternalResponse::Get(response_data, response_token)
+            ),
+        };
+
+        match SignedMessage::new(Address::Node(self.core.id().name()),
+            routing_message, self.core.id().signing_private_key()) {
+                Ok(signed_message) => ignore(self.send(signed_message)),
+                Err(e) => return Err(RoutingError::Cbor(e)),
+            };
+        Ok(())
+    }
+
+    fn handle_cache_put(&mut self, request : ExternalRequest) -> RoutingResult {
```

Comment:
This should only try and put to cache when the messagetype is a get_response, so if a cache is there it will add it, otherwise ignore it. I think this function requires no return.